common/src/main/java/com/wynntils/features/user/SpellCastRenderFeature.java

```diff
@@ -0,0 +1,113 @@
+import com.wynntils.core.components.Models;
+import com.wynntils.core.config.Config;
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.handlers.item.event.ItemRenamedEvent;
+import com.wynntils.mc.event.RenderEvent;
+import com.wynntils.mc.event.TickEvent;
+import com.wynntils.models.items.items.game.GearItem;
+import com.wynntils.models.spells.event.SpellEvent;
+import com.wynntils.utils.colors.CustomColor;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.render.RenderUtils;
+import java.util.Optional;
+import net.minecraft.network.chat.Component;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class SpellCastRenderFeature extends UserFeature {
+    private static final int SHOW_TICKS = 40;
+    public static final int FADE_TICKS = 4;
+
+    @Config
+    public boolean renderVignette = true;
+
+    @Config
+    public float vignetteIntensity = 1.3f;
+
+    @Config
+    public CustomColor vignetteColor = new CustomColor(0, 0, 255);
+
+    private int spellTimer;
+    private Component spellMessage;
+    private float intensity;
+
+    @SubscribeEvent
+    public void onItemRename(ItemRenamedEvent event) {
+        ItemStack itemStack = event.getItemStack();
+        Optional<GearItem> gearItemOpt = Models.Item.asWynnItem(itemStack, GearItem.class);
+        if (gearItemOpt.isEmpty()) return;
+
+        GearItem gearItem = gearItemOpt.get();
+        if (!gearItem.getGearProfile().getGearInfo().getType().isWeapon()) return;
+
+        // Hide vanilla item rename popup
+        event.setCanceled(true);
+    }
+
+    @SubscribeEvent
+    public void onSpellCast(SpellEvent.Cast event) {
+        int manaCost = event.getManaCost();
+        String msg = "§7" + event.getSpellType().getName() + " spell cast! §3[§b-" + manaCost + " ✺§3]";
+        spellMessage = Component.literal(msg);
+
+        // An relativeCost of 1.0 means we just used all mana we have left
+        float relativeCost = (float) manaCost / Models.Character.getCurrentMana();
+        intensity = vignetteIntensity * relativeCost;
+        spellTimer = SHOW_TICKS;
+    }
+
+    @SubscribeEvent
+    public void onSpellFailed(SpellEvent.Failed event) {
+        spellMessage = Component.literal(event.getFailureReason().getMessage());
+        intensity = 0f;
+        spellTimer = SHOW_TICKS;
+    }
+
+    @SubscribeEvent
+    public void onTick(TickEvent event) {
+        if (spellTimer <= 0) return;
+
+        spellTimer--;
+    }
+
+    @SubscribeEvent
+    public void onRender(RenderEvent.Post event) {
+        if (spellTimer <= 0) return;
+
+        if (renderVignette && intensity > 0f) {
+            int shownTicks = SHOW_TICKS - spellTimer;
+            int fade = FADE_TICKS - shownTicks;
+            if (fade > 0) {
+                float alpha = intensity * ((float) fade / FADE_TICKS);
+                RenderUtils.renderVignetteOverlay(event.getPoseStack(), vignetteColor, alpha);
```

Comment:
How does this look like with low health vignette and this?

Reply:
I wondered that myself, but I'm honestly not quite sure how to test that. I need to have a character that has a reasonable amount of spells unlocked, go to some place where I can get a lot of damage, not die, and ...

*duh*

You know what. I'll just hack the low health vignette to start appearing at 95% or something and test. :-) I'll be back.

Reply:
It looks great! It gets kind of purplish, but the impression you get is that the red gets a blue flash. I captured a screen recording but it is too large to upload. But here is a still, from where both blue and red vignette is at max.

<img width="769" alt="Screen Shot 2023-01-25 at 19 25 33" src="https://user-images.githubusercontent.com/6209108/214650439-6514460e-d90b-4f62-a201-63cf96f5ca4d.png">